src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -299,8 +340,39 @@ static void AppendHashtableContents(IndentedTextWriter writer, Hashtable ht)
             }
         }
 
-        /// <summary>Emits the body of the FindFirstChar override.</summary>
-        private static RequiredHelperFunctions EmitFindFirstChar(IndentedTextWriter writer, RegexMethod rm, string id)
+        private static void EmitScan(IndentedTextWriter writer, RegexMethod rm, string id)
+        {
+            using (EmitBlock(writer, "while (true)"))
+            {
+                using (EmitBlock(writer, "if (FindNextPossibleStartingPosition(text))"))
+                {
+                    if (rm.MatchTimeout != Timeout.Infinite)
+                    {
+                        writer.WriteLine("base.CheckTimeout();");
+                        writer.WriteLine();
+                    }
+
+                    writer.WriteLine("// If we got a match, we're done.");
+                    using (EmitBlock(writer, "if (ValidateCurrentPositionMatches(text))"))
+                    {
+                        writer.WriteLine("return;");
+                    }
+                }
+                writer.WriteLine();
+
+                writer.WriteLine("// We failed to find a match. If we're at the end of the input, then we're done.");
+                using (EmitBlock(writer, "if (base.runtextpos == text.Length)"))
+                {
+                    writer.WriteLine("return;");
+                }
+                writer.WriteLine();
+
+                writer.WriteLine("base.runtextpos++;");
```

Comment:
Is this logic right?  In what situation does FindNextPossibleStartingPosition return false but it still needs to loop around again?  I would have expected the logic to be:
```C#
while (FindNextPossibleStartingPosition(text))
{
    if (ValidateCurrentPositionMatches(text) ||
        base.runtextpos == text.Length)
    {
        return;
    }

    base.runtextpos++;
}
```